lib/rack/session/dalli.rb

```diff
@@ -78,95 +66,116 @@ def initialize(app, options = {})
 
-      def get_session(_env, sid)
-        with_block([nil, {}]) do |dc|
-          unless sid && !sid.empty? && (session = dc.get(sid))
-            old_sid = sid
-            sid = generate_sid_with(dc)
-            session = {}
-            unless dc.add(sid, session, @default_ttl)
-              sid = old_sid
-              redo # generate a new sid and try again
-            end
-          end
-          [sid, session]
+      def find_session(_req, sid)
+        with_dalli_client([nil, {}]) do |dc|
+          existing_session = existing_session_for_sid(dc, sid)
+          return [sid, existing_session] unless existing_session.nil?
+
+          [create_sid_with_empty_session(dc), {}]
         end
       end
 
-      def set_session(_env, session_id, new_session, options)
-        return false unless session_id
+      def write_session(_req, sid, session, options)
+        return false unless sid
 
-        with_block(false) do |dc|
-          dc.set(session_id, new_session, ttl(options[:expire_after]))
-          session_id
+        with_dalli_client(false) do |dc|
+          dc.set(memcached_key_from_sid(sid), session, ttl(options[:expire_after]))
+          sid
         end
       end
 
-      def destroy_session(_env, session_id, options)
-        with_block do |dc|
-          dc.delete(session_id)
+      def delete_session(_req, sid, options)
+        with_dalli_client do |dc|
+          dc.delete(memcached_key_from_sid(sid))
           generate_sid_with(dc) unless options[:drop]
         end
       end
 
-      def find_session(req, sid)
-        get_session req.env, sid
+      private
+
+      def memcached_key_from_sid(sid)
+        sid.private_id
       end
 
-      def write_session(req, sid, session, options)
-        set_session req.env, sid, session, options
+      def existing_session_for_sid(client, sid)
+        return nil unless sid && !sid.empty?
+
+        client.get(memcached_key_from_sid(sid))
       end
 
-      def delete_session(req, sid, options)
-        destroy_session req.env, sid, options
+      def create_sid_with_empty_session(client)
+        loop do
+          sid = generate_sid_with(client)
+
+          break sid if client.add(memcached_key_from_sid(sid), {}, @default_ttl)
+        end
       end
 
-      private
+      def generate_sid_with(client)
+        loop do
+          raw_sid = generate_sid
+          sid = raw_sid.is_a?(String) ? Rack::Session::SessionId.new(raw_sid) : raw_sid
+          break sid unless client.get(memcached_key_from_sid(sid))
+        end
+      end
+
+      def build_data_source(options)
+        server_configurations, client_options, pool_options = extract_dalli_options(options)
+
+        if pool_options.empty?
+          ::Dalli::Client.new(server_configurations, client_options)
+        else
+          ensure_connection_pool_added!
+          ConnectionPool.new(pool_options) do
+            ::Dalli::Client.new(server_configurations, client_options.merge(threadsafe: false))
```

Comment:
I can't tell why the args here to changed to be always `threadsafe: false` in a connection pool

to

```ruby          ConnectionPool.new(pool_options) do
            ::Dalli::Client.new(server_configurations, client_options.merge(threadsafe: false))
```

from

```ruby
        mserv, mopts, popts = extract_dalli_options(options)
# where essentially has
#        if mopts[:pool_size] || mopts[:pool_timeout]
#          mopts[:threadsafe] = true
#        end

        @pool = ConnectionPool.new(popts || {}) { ::Dalli::Client.new(mserv, mopts) }
```

context: we have this setup and seeing (sometimes, rarely, in spurts) cache reads get the wrong data

Reply:
@bf4 The whole notion of using a connection pool is that individual threads check out a connection and then check it back in when they are done with it.  While it is checked out it is unavailable for use by any other thread.  Ergo, any thread safety at the `Dalli::Client` is superfluous.

This is a relatively standard usage and is unlikely to be the source of your data corruption issues.  If you can produce a reproducible case I'll take a look.

Reply:
Thanks @petergoldstein for the quick response.   I figure at least now the reason for the change is sufficiently commented :)

I don't expect to have a reproducible case at this point.  I've done quite a bit of logging and this was my shot in the dark.  Guess it's either gremlins or some unreported bug at redislabs or some network shenanigans at Heroku .